### PR TITLE
[SYCL]Use llvm-link's only-needed option to link device libs

### PR DESCRIPTION
**Pull request by jinge90**
Signed-off-by: gejin <ge.jin@intel.com>

This PR aims to solve the same problem in a different way from https://github.com/intel/llvm/pull/2723 which is a better way.
We will do 2-step link:
1. normal llvm-link for all user LLVM bc to generate a "full" user.bc
2. llvm-link -only-needed to link "full" user.bc with device libraries.
In step, we expect only functions really required by user code will be linked in.
We create the PR to run pre-ci tests and CPU performance tests.

### CODE REVIEW

**Comment by jinge90** [comment]
Hi, @bader 
This PR is based on your suggestion "option 2", we split the device link into 2 steps:
1. normal llvm-link for all user's bc
2. llvm-link only needed for devicelib bc.
The code change is little, I ran full test suite and all passed. Also verified some perf benchmarks on CPU device, the perf improvement is as expected. Based on the IR dumped, no "double" functions were introduced into final device image if user code didn't include any double function, I will try on real platforms without double support later.
I will have a look into your "option 1" later.
Thanks very much.

**Comment by bader** [comment]
@jinge90, thanks. Let me know the results of your investigation. I suggest closing #2723 - this solution is preferable in my opinion.

**Comment by jinge90** [comment]
> @jinge90, thanks. Let me know the results of your investigation. I suggest closing #2723 - this solution is preferable in my opinion.

Yes, I will close #2723, no matter option 1 works or not, this PR is much better than #2723.
Thanks very much.

**Review by mdtoguchi — COMMENTED** [review]
Could a test be added to verify when -only-needed is added?

**Comment by jinge90** [comment]
> Could a test be added to verify when -only-needed is added?

Hi, @mdtoguchi 
We need  some more investigation before making final decisions to optimize devicelib default. Once we decide to use this PR, I will add driver test to verify "-only-needed".
Thanks very much.

**Comment by jinge90** [comment]
> Could a test be added to verify when -only-needed is added?

Hi, @mdtoguchi 
I have updated clang driver test sycl-device-lib and sycl-device-lib-win to add check for "llvm-link -only-needed" and I have updated both tests to adapt to latest compiler and device libraries(e.g. libsycl-glibc/msvc.o have been renamed to libsycl-crt.o).
Thanks very much.

**Comment by jinge90** [comment]
Hi, @bader @vzakhari @mdtoguchi 
Could you help to review this patch?
Thanks very much.

File: clang/test/Driver/sycl-device-lib-win.cpp [hunk 1]
```diff
@@ -1,7 +1,7 @@
 ///
 /// Perform several driver tests for SYCL device libraries on Windows
 ///
-// REQUIRES: clang-driver, windows
+// UNSUPPORTED: system-linux
```
Comment (bader) [inline clang/test/Driver/sycl-device-lib-win.cpp:4]: Deleted code seems to express the limitation more precisely. Is it supposed to work on Darwin, Fuchsia or FreeBSD?

File: clang/lib/Driver/ToolChains/SYCL.cpp [hunk 2]
```diff
@@ -101,6 +100,18 @@ const char *SYCL::Linker::constructLLVMLinkCommand(Compilation &C,
...
+  for (const auto &II : InputFiles) {
+    StringRef InputFilename =
+        llvm::sys::path::filename(StringRef(II.getFilename()));
+    StringRef InputSuffix = ".o";
+    if (InputFilename.startswith("libsycl-") &&
+        InputFilename.endswith(InputSuffix)) {
+      LinkSYCLDeviceLibs = true;
+      break;
+    }
+  }
```
Comment (bader) [inline clang/lib/Driver/ToolChains/SYCL.cpp:114]: I think it's better to move this code under if statement at line 120 and replace for-loop with standard library function call e.g. `std::find_if`.

File: clang/lib/Driver/ToolChains/SYCL.cpp [hunk 3]
```diff
@@ -100,24 +100,24 @@ const char *SYCL::Linker::constructLLVMLinkCommand(Compilation &C,
...
+    bool LinkSYCLDeviceLibs = false;
+    auto SYCLDeviceLibIter =
+        std::find_if(InputFiles.begin(), InputFiles.end(), [](const auto &II) {
+          StringRef InputFilename =
+              llvm::sys::path::filename(StringRef(II.getFilename()));
+          if (InputFilename.startswith("libsycl-") &&
+              InputFilename.endswith(".o"))
+            return true;
+          else
+            return false;
```
Comment (bader) [inline clang/lib/Driver/ToolChains/SYCL.cpp:118]:
```suggestion
          return false;
```

File: clang/lib/Driver/ToolChains/SYCL.cpp [hunk 4]
```diff
@@ -100,24 +100,24 @@ const char *SYCL::Linker::constructLLVMLinkCommand(Compilation &C,
...
-      break;
-    }
-  }
   // Add the input bc's created by compile step.
   // When offloading, the input file(s) could be from unbundled partially
   // linked archives.  The unbundled information is a list of files and not
   // an actual object/archive.  Take that list and pass those to the linker
   // instead of the original object.
   if (JA.isDeviceOffloading(Action::OFK_SYCL)) {
+    bool LinkSYCLDeviceLibs = false;
```
Comment (bader) [inline clang/lib/Driver/ToolChains/SYCL.cpp:109]:
```suggestion
```

File: clang/lib/Driver/ToolChains/SYCL.cpp [hunk 5]
```diff
@@ -100,24 +100,24 @@ const char *SYCL::Linker::constructLLVMLinkCommand(Compilation &C,
...
+        std::find_if(InputFiles.begin(), InputFiles.end(), [](const auto &II) {
+          StringRef InputFilename =
+              llvm::sys::path::filename(StringRef(II.getFilename()));
+          if (InputFilename.startswith("libsycl-") &&
+              InputFilename.endswith(".o"))
+            return true;
+          else
+            return false;
+        });
+    LinkSYCLDeviceLibs = (SYCLDeviceLibIter != InputFiles.end());
```
Comment (bader) [inline clang/lib/Driver/ToolChains/SYCL.cpp:120]:
```suggestion
    bool LinkSYCLDeviceLibs = (SYCLDeviceLibIter != InputFiles.end());
```

File: clang/test/Driver/sycl-device-lib-win.cpp [hunk 6]
```diff
@@ -91,3 +91,23 @@
...
+// SYCL_LLVM_LINK_DEVICE_LIB-NEXT: clang-offload-bundler{{.*}} "-type=o" "-targets=sycl-spir64-unknown-unknown-sycldevice" "-inputs={{.*}}libsycl-fallback-cassert.obj" "-outputs={{.*}}libsycl-fallback-cassert-{{.*}}.o" "-unbundle"
+// SYCL_LLVM_LINK_DEVICE_LIB-NEXT: clang-offload-bundler{{.*}} "-type=o" "-targets=sycl-spir64-unknown-unknown-sycldevice" "-inputs={{.*}}libsycl-fallback-complex.obj" "-outputs={{.*}}libsycl-fallback-complex-{{.*}}.o" "-unbundle"
+// SYCL_LLVM_LINK_DEVICE_LIB-NEXT: clang-offload-bundler{{.*}} "-type=o" "-targets=sycl-spir64-unknown-unknown-sycldevice" "-inputs={{.*}}libsycl-fallback-cmath.obj" "-outputs={{.*}}libsycl-fallback-cmath-{{.*}}.o" "-unbundle"
+// SYCL_LLVM_LINK_DEVICE_LIB-NEXT: llvm-link{{.*}} "-only-needed" "{{.*}}" "-o" "{{.*}}.bc" "--suppress-warnings"
+
+/// ###########################################################################
+/// test llvm-link behavior for fno-sycl-device-lib
+// RUN: %clangxx -fsycl -fno-sycl-device-lib=all %s -### 2>&1 \
+// RUN:   | FileCheck %s -check-prefix=SYCL_LLVM_LINK_NO_DEVICE_LIB
+// SYCL_LLVM_LINK_NO_DEVICE_LIB: llvm-link{{.*}}  "{{.*}}.bc" "-o" "{{.*}}.bc" "--suppress-warnings"
```
Comment (mdtoguchi) [inline clang/test/Driver/sycl-device-lib-win.cpp:112]: I would expect that `-only-needed` is not passed in this case - can a negative check be added?

File: clang/test/Driver/sycl-device-lib-win.cpp (hunk 1 above)
Comment (jinge90) [reply clang/test/Driver/sycl-device-lib-win.cpp]: Hi, @bader 
You are correct, this lit test should be only for windows.
Thanks very much.

File: clang/test/Driver/sycl-device-lib.cpp [hunk 7]
```diff
@@ -91,3 +92,24 @@
...
+// SYCL_LLVM_LINK_DEVICE_LIB-NEXT: clang-offload-bundler{{.*}} "-type=o" "-targets=sycl-spir64-unknown-unknown-sycldevice" "-inputs={{.*}}libsycl-fallback-cassert.o" "-outputs={{.*}}libsycl-fallback-cassert-{{.*}}.o" "-unbundle"
+// SYCL_LLVM_LINK_DEVICE_LIB-NEXT: clang-offload-bundler{{.*}} "-type=o" "-targets=sycl-spir64-unknown-unknown-sycldevice" "-inputs={{.*}}libsycl-fallback-complex.o" "-outputs={{.*}}libsycl-fallback-complex-{{.*}}.o" "-unbundle"
+// SYCL_LLVM_LINK_DEVICE_LIB-NEXT: clang-offload-bundler{{.*}} "-type=o" "-targets=sycl-spir64-unknown-unknown-sycldevice" "-inputs={{.*}}libsycl-fallback-cmath.o" "-outputs={{.*}}libsycl-fallback-cmath-{{.*}}.o" "-unbundle"
+// SYCL_LLVM_LINK_DEVICE_LIB-NEXT: llvm-link{{.*}} "-only-needed" "{{.*}}" "-o" "{{.*}}.bc" "--suppress-warnings"
+
+/// ###########################################################################
+/// test llvm-link behavior for fno-sycl-device-lib
+// RUN: %clangxx -fsycl -fno-sycl-device-lib=all %s -### 2>&1 \
+// RUN:   | FileCheck %s -check-prefix=SYCL_LLVM_LINK_NO_DEVICE_LIB
+// SYCL_LLVM_LINK_NO_DEVICE_LIB: clang{{.*}} "-cc1" {{.*}} "-fsycl-is-device"
```
Comment (jinge90) [inline clang/test/Driver/sycl-device-lib.cpp:113]: @mdtoguchi 
I have updated the test case to check if "only-needed" is excluded when "-fno-sycl-device-lib=all" is used.
Thanks very much.

**Review by mdtoguchi — APPROVED** [review]
LGTM

File: clang/lib/Driver/ToolChains/SYCL.cpp [hunk 8]
```diff
@@ -107,8 +106,20 @@ const char *SYCL::Linker::constructLLVMLinkCommand(Compilation &C,
...
   // instead of the original object.
   if (JA.isDeviceOffloading(Action::OFK_SYCL)) {
+    auto SYCLDeviceLibIter =
+        std::find_if(InputFiles.begin(), InputFiles.end(), [](const auto &II) {
+          StringRef InputFilename =
+              llvm::sys::path::filename(StringRef(II.getFilename()));
+          if (InputFilename.startswith("libsycl-") &&
+              InputFilename.endswith(".o"))
+            return true;
+          return false;
```
Comment (sndmitriev) [inline clang/lib/Driver/ToolChains/SYCL.cpp:116]: Using file name for deciding if `--only-needed` option is required looks very fragile.

Comment (jinge90) [reply clang/lib/Driver/ToolChains/SYCL.cpp:116]: Hi, @sndmitriev 
I agree, we need some enhancement here .

**Pull request merged**